core/typeSugar.ml

```diff
@@ -4898,12 +5127,27 @@ and type_binding : context -> binding -> binding * context * Usage.t =
           let () = unify pos ~handle:Gripers.bind_exp
             (pos_and_typ e, no_pos Types.unit_type) in
           Exp (erase e), empty_context, usages e
+      | FreshLabel ls ->
```

Comment:
I don't understand what's happening here. It seems that at this point, some local labels `ls` are brought into scope, but then we try to remove occurences of these labels from types in `context`, which is the typing context before `ls` are brought into scope?

Is similar logic necessary for when labels are brought into scope using the `fresh` keyword in, say, a function definition? I assume the code here is only run for standalone `fresh` declarations, right?

Reply:
I think there is a bug here actually. The labels are bought into to scope, but I think they should be removed when the block is **exited**. For this to work correctly, I think we need to track the "level" (i.e. block nesting depth) that the labels were introduced at.